File: ObjC/AppLSObjC/AppLSObjC/FileListViewController.m
//
//  FileListViewController.m
//  AppLSObjC
//
//  Created by Qua Zi Xian on 27/4/15.
//  Copyright (c) 2015 ZXLabs. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import "FileListViewController.h"

// Defines an item in a directory.
// name represents the item name.
// isDir indicates it the item is a directory.
@interface Descriptor : NSObject 

@property (nonatomic, readonly) NSString *name;
@property (nonatomic, readonly) BOOL *isDir;

-(id)init:(NSString*)name isDir:(BOOL*)isDir;

@end

@implementation Descriptor: NSObject

-(id)init:(NSString*)name isDir:(BOOL*)isDir {
    if(self = [super init]) {
        _name = name;
        _isDir = isDir;
        return self;
    } else {
        return nil;
    }
}

@end

@implementation FileListViewController: UITableViewController

int _numSections = 1;
NSString *_cellIdentifier = @"fileListCell";
NSString *_dir;
NSArray *_contents;

-(id)init {
    if(self = [super initWithStyle:UITableViewStylePlain]) {
        _dir = (NSString*)[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, true) objectAtIndex:(NSUInteger)0];
        _contents = [self getContents];
        self.title = [_dir lastPathComponent];
        [self navigationItem].rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(rightBarButtonClicked:)];
        [[self tableView] registerClass:[UITableViewCell class] forCellReuseIdentifier:_cellIdentifier];
        [self tableView].dataSource = self;
        [self tableView].delegate = self;
        return self;
    } else {
        return nil;
    }
}

-(id)initForDirectory:(NSString *)directory {
    if(self = [super initWithStyle:UITableViewStylePlain]) {
        _dir = directory;
        return self;
    } else {
        return nil;
    }
}

// UITableViewDataSource methods
-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return _numSections;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [_contents count];
}

-(UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell* cell = [[self tableView] dequeueReusableCellWithIdentifier:_cellIdentifier forIndexPath:indexPath];
    if(cell == nil) {
        cell = [[UITableViewCell alloc] init];
    }
    cell.textLabel.text = [_contents objectAtIndex:[indexPath row]];
    return cell;
}
//End of UITableViewDataSource methods

// Dismisses the app when Cancel button is tapped.
-(void)rightBarButtonClicked:(UIBarButtonItem*)sender {
    [[self navigationController] dismissViewControllerAnimated:true completion:nil];
}

// Gets list of files and folders in the given directory.
-(NSArray*)getContents {
    NSMutableArray *arr = [[NSMutableArray alloc] init];
    NSArray *fileList = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:_dir error:nil];
    
    // Construct the respective descriptor for each item
    for (NSString *file in fileList) {
        BOOL *isDir;
        [[NSFileManager defaultManager] fileExistsAtPath:[_dir stringByAppendingPathComponent:file] isDirectory:isDir];
        [arr addObject:[[Descriptor alloc] init:file isDir:isDir]];
    }
    return arr;
}

@end

File: Swift/AppLS-Swift/AppLS-Swift/FileListViewController.swift
//
//  FileListViewController.swift
//  AppLS-Swift
//
//  The table view controller resposible for listing files and folders in the given directory.
//
//  Created by Qua Zi Xian on 26/4/15.
//  Copyright (c) 2015 ZXLabs. All rights reserved.
//

import Foundation
import UIKit

class FileListViewController: UITableViewController {
    private let numSections = 1
    private let cellIdentifier = "fileListCell"
    private let _dir: String
    private var _contents: [Descriptor]!
    
    init() {
        _dir = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true)[0] as! String
        super.init(nibName: nil, bundle: nil)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.Cancel, target: self, action: "rightButtonClicked:")
        _contents = getContents()
        title = _dir.lastPathComponent
        tableView.delegate = self
        tableView.dataSource = self
        tableView.registerClass(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)
    }
    
    init(directory: String) {
        _dir = directory
        super.init(nibName: nil, bundle: nil)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.Cancel, target: self, action: "rightButtonClicked:")
        _contents = getContents()
        title = _dir.lastPathComponent
        tableView.delegate = self
        tableView.dataSource = self
        tableView.registerClass(UITableViewCell.self, forCellReuseIdentifier: cellIdentifier)
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    //UITableViewDelegate methods
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let row = indexPath.row
        if row >= 0 && row < _contents.count {
            if _contents[row].isDir {
                navigationController?.pushViewController(FileListViewController(directory: _dir.stringByAppendingPathComponent(_contents[row].name)), animated: true)
                tableView.deselectRowAtIndexPath(indexPath, animated: true)
            } else {
                if let destinationVC = FileDescriptorViewController(filePath: _dir.stringByAppendingPathComponent(_contents[row].name)) {
                    navigationController?.pushViewController(destinationVC, animated: true)
                }
                tableView.deselectRowAtIndexPath(indexPath, animated: false)
            }
        }
    }
    //End of UITableViewDelegate methods
    
    //UITableViewDataSource methods
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return numSections
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _contents.count
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(cellIdentifier, forIndexPath: indexPath) as! UITableViewCell
        cell.textLabel?.text = _contents[indexPath.row].name
        if _contents[indexPath.row].isDir {
            cell.accessoryType = UITableViewCellAccessoryType.DisclosureIndicator
        } else {
            cell.accessoryType = UITableViewCellAccessoryType.DetailButton
        }
        return cell
    }
    //End of UITableViewDataSource methods
    
    private func getContents() -> [Descriptor] {
        var isDir: ObjCBool = false
        let fileList = NSFileManager.defaultManager().contentsOfDirectoryAtPath(_dir, error: nil) as! [String]
        var arr = [Descriptor]()
        arr.reserveCapacity(fileList.count)
        
        //Create appropriate descriptor for each item in directory.
        for file in fileList {
            NSFileManager.defaultManager().fileExistsAtPath(_dir.stringByAppendingPathComponent(file), isDirectory: &isDir)
            arr.append(Descriptor(name: file, isDir: Bool(isDir)))
        }
        
        return arr
    }
    
    /// Dismisses the app on tapping right bar button.
    func rightButtonClicked(sender: UIBarButtonItem) {
        navigationController?.dismissViewControllerAnimated(true, completion: nil)
    }
}